### PR TITLE
Dev minimax mark

### PR DESCRIPTION
**Pull request by maxiao124**
Add Minimax supported
#100
Welcome to comment @li2109

### CODE REVIEW

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private static final String BASE_URL = "https://api.minimax.chat";
+
+  private final MiniMaxApi miniMaxApi;
+
+  public static MiniMaxService create(final MiniMaxApi api) {
+    return new MiniMaxService(api);
+  }
+
+  public static MiniMaxService create(final MiniMaxServiceConfig miniMaxServiceConfig) {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java:41]: Sorry. I understand that you are following the old style; however, we are switching to use Guice( a dependency injection framework) to decouple components. Please take a look at https://github.com/Knowly-ai/langtorch/issues/119

This means we would use less static factory method to do instance initialization (we had a lot of this before the switch) but going for default constructor that has @Inject annotation. Please update the code using latest OpenAIService as a reference. Thank you!!

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  public ListenableFuture<ChatCompletionResult> createChatCompletionAsync(
+      ChatCompletionRequest request) {
+    return miniMaxApi.createChatCompletion(request);
+  }
+
+  public EmbeddingResult createEmbeddings(EmbeddingRequest request) {
+    return execute(miniMaxApi.createEmbeddings(request));
+  }
+
+  public ListenableFuture<EmbeddingResult> createEmbeddingsAsync(EmbeddingRequest request) {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java:122]: Please add retry

Example:
https://github.com/Knowly-ai/langtorch/blob/55f9b370a883e5a40bc20eaab51f1fee56fcdfb0/src/main/java/ai/knowly/langtorch/llm/openai/OpenAIService.java#L182

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    try {
+      return apiCall.get();
+    } catch (InterruptedException e) {
+      // Restore the interrupt status
+      Thread.currentThread().interrupt();
+      // Optionally, log or handle the exception here.
+      logger.atSevere().withCause(e).log("Thread was interrupted during API call.");
+      throw new MiniMaxApiServiceInterruptedException(e);
+
+    } catch (ExecutionException e) {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java:139]: The actual exception is buried deeply inside the cause where you can get from e.getCause() and specifically it's a httpException
https://github.com/square/retrofit/blob/69315219bc9068b82a3c320fd6297fed045ece7e/retrofit-adapters/guava/src/main/java/retrofit2/adapter/guava/GuavaCallAdapterFactory.java#L48

Please update here according based on:
https://github.com/Knowly-ai/langtorch/blob/55f9b370a883e5a40bc20eaab51f1fee56fcdfb0/src/main/java/ai/knowly/langtorch/llm/openai/OpenAIService.java#LL96C7-L107C8

and please also write a test for this.

File: src/main/java/ai/knowly/langtorch/llm/minimax/schema/config/MiniMaxServiceConfig.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.google.auto.value.AutoValue;
+import java.time.Duration;
+
+/**
+ * @author maxiao
+ * @date 2023/06/07
+ */
+@AutoValue
+public abstract class MiniMaxServiceConfig {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/llm/minimax/schema/config/MiniMaxServiceConfig.java:11]: If you are adding the retry, you probably need to update this too.

example:
https://github.com/Knowly-ai/langtorch/blob/55f9b370a883e5a40bc20eaab51f1fee56fcdfb0/src/main/java/ai/knowly/langtorch/llm/openai/schema/config/OpenAIServiceConfig.java#L11

File: src/main/java/ai/knowly/langtorch/processor/minimax/MiniMaxServiceProvider.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import ai.knowly.langtorch.utils.Environment;
+import ai.knowly.langtorch.utils.api.key.MiniMaxKeyUtil;
+import com.google.common.flogger.FluentLogger;
+import java.time.Duration;
+
+/**
+ * @author maxiao
+ * @date 2023/06/07
+ */
+public final class MiniMaxServiceProvider {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/processor/minimax/MiniMaxServiceProvider.java:14]: Could you replace this with Guice module? 

example:
https://github.com/Knowly-ai/langtorch/blob/55f9b370a883e5a40bc20eaab51f1fee56fcdfb0/src/main/java/ai/knowly/langtorch/llm/openai/OpenAIServiceConfigWithExplicitAPIKeyModule.java#L8

File: src/main/java/ai/knowly/langtorch/processor/minimax/chat/MiniMaxChatProcessor.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  private MiniMaxChatProcessorConfig miniMaxChatProcessorConfig =
+      MiniMaxChatProcessorConfig.builder().setModel(DEFAULT_MODEL).build();
+
+  // Constructor with dependency injection
+  @Inject
+  MiniMaxChatProcessor(MiniMaxService miniMaxService) {
+    this.miniMaxService = miniMaxService;
+  }
+
+  // Private constructor used in factory methods
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/processor/minimax/chat/MiniMaxChatProcessor.java:39]: Let's put everything we need into the constructor(make it public) so that injection would work

example:
OpenAITextProcessor

File: src/main/java/ai/knowly/langtorch/processor/minimax/embeddings/MiniMaxEmbeddingsProcessorRequestConverter.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+
+public final class MiniMaxEmbeddingsProcessorRequestConverter {
+  private MiniMaxEmbeddingsProcessorRequestConverter() {}
+
+  public static EmbeddingRequest convert(
+      MiniMaxEmbeddingsProcessorConfig MiniMaxEmbeddingsProcessorConfig,
+      String model,
+      List<String> texts,
+      String type) {
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/processor/minimax/embeddings/MiniMaxEmbeddingsProcessorRequestConverter.java:13]: Does it make sense to make an enum for type?

File: src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java [hunk 8]
```diff
@@ -8,15 +8,26 @@ public class Embedding implements Output {
...
+  private final List<Float> floatVector;
+
+  private Embedding(List<Double> vector, List<Float> floatVector) {
     this.vector = vector;
+    this.floatVector = floatVector;
   }
 
   public static Embedding of(List<Double> vector) {
-    return new Embedding(vector);
+    return new Embedding(vector, null);
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java:19]: why not a empty list?

File: src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java [hunk 9]
```diff
@@ -8,15 +8,26 @@ public class Embedding implements Output {
...
+  }
+
+  public static Embedding ofFloatVector(List<Float> floatVector) {
+    return new Embedding(null, floatVector);
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java:23]: empty list?

File: src/main/java/ai/knowly/langtorch/utils/api/key/OpenAIKeyUtil.java [hunk 10]
```diff
@@ -7,7 +7,7 @@
 import com.google.common.flogger.FluentLogger;
 import java.util.Optional;
 
-/** Get Pinecone key from .env file */
```
Comment (li2109) [inline src/main/java/ai/knowly/langtorch/utils/api/key/OpenAIKeyUtil.java:10]: nice catch!!

File: src/test/java/ai/knowly/langtorch/llm/minimax/ChatCompletionTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        ChatCompletionRequest.Message.builder()
+            .setSenderType("USER")
+            .setText("Hello, how are you?")
+            .build());
+
+    ChatCompletionRequest chatCompletionRequest =
+        ChatCompletionRequest.builder().setModel("abab5-chat").setMessages(messages).build();
+
+    String reply = service.createChatCompletion(chatCompletionRequest).getReply();
+    assertNotNull(reply);
```
Comment (li2109) [inline src/test/java/ai/knowly/langtorch/llm/minimax/ChatCompletionTest.java:29]: please use assertThat from google's truth library so that being consistent with other tests.

**Review by li2109 — CHANGES_REQUESTED** [review]
Wow! Thank you so much for the PR! it must require a lot of work!

Generally the PR looks good and there are some nits though.
I want to say sorry about the style change that's due to https://github.com/Knowly-ai/langtorch/issues/119

Because of this, you shall see many changes in the code where we used to have many static factory methods but now we dont. It makes no sense to have that if we are initializing a class by injection.

**Comment by li2109** [comment]
BTW, the PR does not build though.

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxApi.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import ai.knowly.langtorch.llm.minimax.schema.dto.completion.ChatCompletionRequest;
+import ai.knowly.langtorch.llm.minimax.schema.dto.completion.ChatCompletionResult;
+import ai.knowly.langtorch.llm.minimax.schema.dto.embedding.EmbeddingRequest;
+import ai.knowly.langtorch.llm.minimax.schema.dto.embedding.EmbeddingResult;
+import com.google.common.util.concurrent.ListenableFuture;
+import retrofit2.http.Body;
+import retrofit2.http.POST;
+
+/**
+ * doc link: https://api.minimax.chat/document/guides?id=6433f37194878d408fc8294b
```
[the remark this answers is not in the document]
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxApi.java]: had done

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java (hunk 1 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java]: It's pretty a good idea, the code is more readability

had done this

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java (hunk 2 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java:114]: had done

File: src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java (hunk 3 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/llm/minimax/MiniMaxService.java:138]: I missed here，thank you.Minimax does not directly throw an HTTP exception, but carries an error code in the returned body. So I resoled this issue outside.

`
  public ChatCompletionResult createChatCompletion(ChatCompletionRequest request) {
    ChatCompletionResult chatCompletionResult =
        execute(
            futureRetrier.runWithRetries(() -> api.createChatCompletion(request), result -> true));

    checkResp(chatCompletionResult.getBaseResp());
    return chatCompletionResult;
  }

  /** Throw exception messages if the request fails */
  public void checkResp(BaseResp baseResp) {
    if (baseResp.getStatusCode() != 0) {
      throw new MiniMaxApiBusinessErrorException(baseResp.getStatusCode(), baseResp.getStatusMsg());
    }
  }
`

File: src/main/java/ai/knowly/langtorch/llm/minimax/schema/config/MiniMaxServiceConfig.java (hunk 4 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/llm/minimax/schema/config/MiniMaxServiceConfig.java:14]: had done

File: src/main/java/ai/knowly/langtorch/processor/minimax/MiniMaxServiceProvider.java (hunk 5 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/processor/minimax/MiniMaxServiceProvider.java:13]: had done

File: src/main/java/ai/knowly/langtorch/processor/minimax/chat/MiniMaxChatProcessor.java (hunk 6 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/processor/minimax/chat/MiniMaxChatProcessor.java]: had done

File: src/main/java/ai/knowly/langtorch/processor/minimax/embeddings/MiniMaxEmbeddingsProcessorRequestConverter.java (hunk 7 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/processor/minimax/embeddings/MiniMaxEmbeddingsProcessorRequestConverter.java:13]: yeah，add class <MiniMaxEmbeddingTypeScene>

File: src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java (hunk 8 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java]: had done

File: src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java (hunk 9 above)
Comment (maxiao124) [reply src/main/java/ai/knowly/langtorch/schema/embeddings/Embedding.java]: had done

File: src/test/java/ai/knowly/langtorch/llm/minimax/ChatCompletionTest.java (hunk 11 above)
Comment (maxiao124) [reply src/test/java/ai/knowly/langtorch/llm/minimax/ChatCompletionTest.java]: had done

**Comment by li2109** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=Knowly-ai_langtorch&pullRequest=132)
> 
> [Bug](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=CODE_SMELL) [10 Code Smells](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=132&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=132&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=132&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=132&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=132&metric=new_duplicated_lines_density&view=list)

Could you resolve the code smell warning?

**Review by li2109 — CHANGES_REQUESTED** [review]
Please resolve the code smell warning
